In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
!git clone https://github.com/Project-MONAI/MONAI.git

Cloning into 'MONAI'...
remote: Enumerating objects: 36913, done.
remote: Counting objects: 100% (159/159), done.
remote: Compressing objects: 100% (114/114), done.
remote: Total 36913 (delta 68), reused 109 (delta 45), pack-reused 36754
Receiving objects: 100% (36913/36913), 64.09 MiB | 12.44 MiB/s, done.
Resolving deltas: 100% (29628/29628), done.


In [ ]:
%cd MONAI/


/content/MONAI


In [ ]:
!pip install -e '.[all]'

Obtaining file:///content/MONAI
  Installing build dependencies ... done
  Checking if build backend supports build_editable ... done
  Getting requirements to build editable ... done
  Preparing editable metadata (pyproject.toml) ... done
  Using cached ninja-1.11.1.1-py2.py3-none-manylinux1_x86_64.manylinux_2_5_x86_64.whl (307 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 266.5/266.5 kB 5.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 299.2/299.2 kB 28.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 9.7/9.7 MB 70.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 316.6/316.6 kB 36.6 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 37.3/37.3 MB 43.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 44.6/44.6 kB 6.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 19.1/19.1 MB 88.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.1/1.1 M

In [ ]:
model_dir = '/content/drive/MyDrive/template/our-git-project/Cross_Val_Results_final_3'

In [ ]:
import os
import torch
import numpy as np
from tqdm import tqdm
import matplotlib.pyplot as plt
from collections import OrderedDict

from monai.apps import download_url
from monai.utils import set_determinism
from monai.losses import DiceCELoss
from monai.inferers import sliding_window_inference
from monai.config import print_config
from monai.transforms import (
    AsDiscrete,
    Compose,
    CropForegroundd,
    EnsureChannelFirstd,
    LoadImaged,
    Orientationd,
    RandFlipd,
    RandCropByPosNegLabeld,
    RandShiftIntensityd,
    ScaleIntensityRanged,
    Spacingd,
    RandRotate90d,
    ToTensord,
)

from monai.metrics import DiceMetric
from monai.networks.nets import UNETR

from monai.data import (
    DataLoader,
    CacheDataset,
    load_decathlon_datalist,
    decollate_batch,
)

print_config()

MONAI version: 1.3.0+42.gd5585c34
Numpy version: 1.23.5
Pytorch version: 2.1.0+cu118
MONAI flags: HAS_EXT = False, USE_COMPILED = False, USE_META_DICT = False
MONAI rev id: d5585c34fc44f46af9ffb28522ce44c7ed869654
MONAI __file__: /content/MONAI/monai/__init__.py

Optional dependencies:
Pytorch Ignite version: 0.4.11
ITK version: 5.3.0
Nibabel version: 4.0.2
scikit-image version: 0.19.3
scipy version: 1.11.4
Pillow version: 9.4.0
Tensorboard version: 2.14.1
gdown version: 4.6.6
TorchVision version: 0.16.0+cu118
tqdm version: 4.66.1
lmdb version: 1.4.1
psutil version: 5.9.5
pandas version: 1.5.3
einops version: 0.7.0
transformers version: 4.21.3
mlflow version: 2.9.0
pynrrd version: 1.0.0
clearml version: 1.13.2

For details about installing the optional dependencies, please visit:
    https://docs.monai.io/en/latest/installation.html#installing-the-recommended-dependencies



In [ ]:
set_determinism(seed=960)

In [ ]:
use_pretrained = True


if use_pretrained:

    dst = '/content/drive/MyDrive/template/our-git-project/Pretrained_Results_final_v2/best_model.pt'

    pretrained_path = os.path.normpath(dst)


In [ ]:
import os
from glob import glob
import shutil
from tqdm import tqdm
import numpy as np
import nibabel as nib
from monai.transforms import(
    Compose,
    EnsureChannelFirstd,
    LoadImaged,
    Resized,
    ToTensord,
    Spacingd,
    Orientationd,
    NormalizeIntensityd,
    CropForegroundd,
    RandShiftIntensityd,
    RandFlipd,
    FgBgToIndicesd,
    RandRotate90d,
    RandCropByPosNegLabeld,
    Activations,



)
from monai.data import DataLoader, Dataset, CacheDataset
from monai.utils import set_determinism

# Training Hyper-params
lr = 1e-5
max_iterations = 10000
eval_num = 100

original_data_folder = '/content/drive/MyDrive/template/our-git-project/preprocessed_images/'
mask_data_folder = '/content/drive/MyDrive/template/our-git-project/preprocessed_images/'

# List of image file paths
images  = sorted(glob(os.path.join(original_data_folder, '*_orig.nii.gz')))
segs  = sorted(glob(os.path.join(mask_data_folder, '*_masks.nii.gz')))

data_dicts  = [{"vol": image_name, "seg": label_name} for image_name, label_name in zip(images, segs)]
train_files, test_files = data_dicts[:-25], data_dicts[-25:]


train_transforms = Compose(

[
            LoadImaged(keys=["vol", "seg"]),
            EnsureChannelFirstd(keys=["vol", "seg"]),
            #Spacingd(keys=["vol", "seg"], pixdim=pixdim, mode=("bilinear", "nearest")),
            Orientationd(keys=["vol", "seg"], axcodes="RAS"),
            #NormalizeIntensityd(keys="vol", nonzero=True, channel_wise=True),
            CropForegroundd(keys=["vol", "seg"], source_key="vol"),
            RandCropByPosNegLabeld(
            keys=["vol", "seg"],
            label_key="seg",
            spatial_size=(64, 64, 64),
            pos=4,
            neg=1,
            num_samples=5,
            image_key="vol",
            image_threshold=0,
            ),
            RandFlipd(
            keys=["vol", "seg"],
            spatial_axis=[0],
            prob=0.1,
        ),
        RandFlipd(
            keys=["vol", "seg"],
            spatial_axis=[1],
            prob=0.1,
        ),
        RandFlipd(
            keys=["vol", "seg"],
            spatial_axis=[2],
            prob=0.1,
        ),
        RandRotate90d(
            keys=["vol", "seg"],
            prob=0.3,
            max_k=3,
            spatial_axes=[0, 2],
        ),
        ToTensord(keys=["vol", "seg"]),

        ]
)
val_transforms = Compose(
        [
            LoadImaged(keys=["vol", "seg"]),
            EnsureChannelFirstd(keys=["vol", "seg"]),
            #Spacingd(keys=["vol", "seg"], pixdim=pixdim, mode=("bilinear", "nearest")),
            Orientationd(keys=["vol", "seg"], axcodes="RAS"),
            #NormalizeIntensityd(keys="vol", nonzero=True, channel_wise=True),
            CropForegroundd(keys=['vol', 'seg'], source_key='vol'),
            RandCropByPosNegLabeld(
            keys=["vol", "seg"],
            label_key="seg",
            spatial_size=(64, 64, 64),
            pos=4,
            neg=1,
            num_samples=5,
            image_key="vol",
            image_threshold=0,
            ),
            RandFlipd(
            keys=["vol", "seg"],
            spatial_axis=[0],
            prob=0.1,
        ),
            RandFlipd(
            keys=["vol", "seg"],
            spatial_axis=[1],
            prob=0.1,
        ),

        RandFlipd(
            keys=["vol", "seg"],
            spatial_axis=[2],
            prob=0.1,
        ),
        RandRotate90d(
            keys=["vol", "seg"],
            prob=0.1,
            max_k=3,
            spatial_axes=[0, 2],
        ),
        ToTensord(keys=["vol", "seg"]),


        ]
)



monai.transforms.croppad.dictionary CropForegroundd.__init__:allow_smaller: Current default value of argument `allow_smaller=True` has been deprecated since version 1.2. It will be changed to `allow_smaller=False` in version 1.5.


In [ ]:
from sklearn.model_selection import KFold
import os
from monai.networks.nets import UNet,UNETR
from monai.networks.layers import Act, Norm
from monai.losses import DiceLoss, DiceCELoss
import torch

# Number of folds
num_folds = 5

# Use sklearn's KFold to split the data into 5 folds
kf = KFold(n_splits=num_folds, shuffle=True, random_state=42)
fold_datasets = list(kf.split(train_files))

# Lists to store results for each fold
fold_results = []

In [ ]:
# Iterate over folds
for fold, (train_indices, val_indices) in enumerate(fold_datasets):
    print(f"\nFold {fold + 1}/{num_folds}")

    if(fold > 1):
        del train_ds, train_loader
        torch.cuda.empty_cache()

        del val_ds, val_loader
        torch.cuda.empty_cache()

    # Create training and validation datasets for the current fold
    train_files = [data_dicts[i] for i in train_indices]
    val_files = [data_dicts[i] for i in val_indices]

    train_ds = CacheDataset(
    data=train_files,
    transform=train_transforms,
    cache_num=24,
    cache_rate=1.0,
    num_workers=4,
    )

    val_ds = CacheDataset(data=val_files, transform=val_transforms, cache_num=6, cache_rate=1.0, num_workers=4)

    # Create data loaders for training and validation sets
    train_loader = DataLoader(train_ds, batch_size=1, shuffle=True, num_workers=4, pin_memory=True)
    val_loader = DataLoader(val_ds, batch_size=1, shuffle=False, num_workers=4, pin_memory=True)

    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
    model = UNETR(
            in_channels=1,
            out_channels=2,
            img_size=(64, 64, 64),
            feature_size=16,
            hidden_size=768,
            mlp_dim=3072,
            num_heads=12,
            pos_embed="conv",
            norm_name="instance",
            res_block=True,
            dropout_rate=0.0,
        ).to(device)

    # Load ViT backbone weights into UNETR
    if use_pretrained is True:
            print("Loading Weights from the Path {}".format(pretrained_path))
            vit_dict = torch.load(pretrained_path)
            vit_weights = vit_dict["state_dict"]

            # Remove items of vit_weights if they are not in the ViT backbone (this is used in UNETR).
            # For example, some variables names like conv3d_transpose.weight, conv3d_transpose.bias,
            # conv3d_transpose_1.weight and conv3d_transpose_1.bias are used to match dimensions
            # while pretraining with ViTAutoEnc and are not a part of ViT backbone.
            model_dict = model.vit.state_dict()

            vit_weights = {k: v for k, v in vit_weights.items() if k in model_dict}
            model_dict.update(vit_weights)
            model.vit.load_state_dict(model_dict)
            del model_dict, vit_weights, vit_dict
            print("Pretrained Weights Succesfully Loaded !")

    elif use_pretrained is False:
            print("No weights were loaded, all weights being used are randomly initialized!")

    model.to(device)

    loss_function = DiceCELoss(to_onehot_y=True, softmax=True,  jaccard=True)
    torch.backends.cudnn.benchmark = True
    optimizer = torch.optim.AdamW(model.parameters(), lr=lr, weight_decay=1e-5)

    post_label = AsDiscrete(to_onehot=2)
    post_pred = AsDiscrete(argmax=True, to_onehot=2)
    dice_metric = DiceMetric(include_background=True, reduction="mean", get_not_nans=False)
    global_step = 0
    dice_val_best = 0.0
    global_step_best = 0
    epoch_loss_values = []
    metric_values = []

    def validation(epoch_iterator_val):
        model.eval()
        dice_vals = []

        with torch.no_grad():
            for _step, batch in enumerate(epoch_iterator_val):
                val_inputs, val_labels = (batch["vol"].cuda(), batch["seg"].cuda())
                val_outputs = sliding_window_inference(val_inputs, (64, 64, 64), 4, model)
                val_labels_list = decollate_batch(val_labels)
                val_labels_convert = [post_label(val_label_tensor) for val_label_tensor in val_labels_list]
                val_outputs_list = decollate_batch(val_outputs)
                val_output_convert = [post_pred(val_pred_tensor) for val_pred_tensor in val_outputs_list]
                dice_metric(y_pred=val_output_convert, y=val_labels_convert)
                dice = dice_metric.aggregate().item()
                dice_vals.append(dice)
                epoch_iterator_val.set_description("Validate (%d / %d Steps) (dice=%2.5f)" % (global_step, 10.0, dice))

            dice_metric.reset()

        mean_dice_val = np.mean(dice_vals)

        return mean_dice_val


    def train(global_step, train_loader, dice_val_best, global_step_best):
        model.train()
        epoch_loss = 0
        step = 0
        epoch_iterator = tqdm(train_loader, desc="Training (X / X Steps) (loss=X.X)", dynamic_ncols=True)
        for step, batch in enumerate(epoch_iterator):
            step += 1
            x, y = (batch["vol"].cuda(), batch["seg"].cuda())
            logit_map = model(x)
            loss = loss_function(logit_map, y)
            loss.backward()
            epoch_loss += loss.item()
            optimizer.step()
            optimizer.zero_grad()
            epoch_iterator.set_description("Training (%d / %d Steps) (loss=%2.5f)" % (global_step, max_iterations, loss))

            if (global_step % eval_num == 0 and global_step != 0) or global_step == max_iterations:
                epoch_iterator_val = tqdm(val_loader, desc="Validate (X / X Steps) (dice=X.X)", dynamic_ncols=True)
                dice_val = validation(epoch_iterator_val)

                epoch_loss /= step
                epoch_loss_values.append(epoch_loss)
                metric_values.append(dice_val)
                if dice_val > dice_val_best:
                    dice_val_best = dice_val
                    global_step_best = global_step
                    torch.save(model.state_dict(), os.path.join(model_dir, "best_metric_model.pth"))
                    print(
                        "Model Was Saved ! Current Best Avg. Dice: {} Current Avg. Dice: {}".format(dice_val_best, dice_val)
                    )
                else:
                    print(
                        "Model Was Not Saved ! Current Best Avg. Dice: {} Current Avg. Dice: {}".format(
                            dice_val_best, dice_val
                        )
                    )

                plt.figure(1, (12, 6))
                plt.subplot(1, 2, 1)
                plt.title("Iteration Average Loss")
                x = [eval_num * (i + 1) for i in range(len(epoch_loss_values))]
                y = epoch_loss_values
                plt.xlabel("Iteration")
                plt.plot(x, y)
                plt.grid()
                plt.subplot(1, 2, 2)
                plt.title("Val Mean Dice")
                x = [eval_num * (i + 1) for i in range(len(metric_values))]
                y = metric_values
                plt.xlabel("Iteration")
                plt.plot(x, y)
                plt.grid()
                plt.savefig(os.path.join(model_dir, "unet_finetune_quick_update.png"))
                plt.clf()
                plt.close(1)

            global_step += 1

        return global_step, dice_val_best, global_step_best


    while global_step < max_iterations:
        global_step, dice_val_best, global_step_best = train(global_step, train_loader, dice_val_best, global_step_best)
    model.load_state_dict(torch.load(os.path.join(model_dir, "best_metric_model.pth")))

    # Print metrics for the current fold
    print(f"Fold {fold + 1}/{num_folds} completed, best_metric: {dice_val_best:.4f} "
          f"at iteration: {global_step_best}")
    #fold_results = fold_results.append(dice_val_best)

# Print overall cross-validation results
print(f"\nOverall cross-validation completed, best_metric: {dice_val_best:.4f} "
      f"at iteration: {global_step_best}")




Fold 1/5


Loading dataset: 100%|██████████| 6/6 [00:04<00:00,  1.39it/s]
monai.networks.nets.unetr UNETR.__init__:pos_embed: Argument `pos_embed` has been deprecated since version 1.2. It will be removed in version 1.4. please use `proj_type` instead.


Loading Weights from the Path /content/drive/MyDrive/template/our-git-project/Pretrained_Results_final_v2/best_model.pt
Pretrained Weights Succesfully Loaded !


Validate (68 / 10 Steps) (dice=0.65176): 100%|██████████| 17/17 [00:07<00:00,  2.31it/s]


Model Was Saved ! Current Best Avg. Dice: 0.6615856991094702 Current Avg. Dice: 0.6615856991094702


Validate (136 / 10 Steps) (dice=0.65715): 100%|██████████| 17/17 [00:07<00:00,  2.31it/s]


Model Was Saved ! Current Best Avg. Dice: 0.6676866727716783 Current Avg. Dice: 0.6676866727716783


Validate (272 / 10 Steps) (dice=0.69398): 100%|██████████| 17/17 [00:07<00:00,  2.18it/s]


Model Was Saved ! Current Best Avg. Dice: 0.715444357956157 Current Avg. Dice: 0.715444357956157


Validate (340 / 10 Steps) (dice=0.71748): 100%|██████████| 17/17 [00:07<00:00,  2.27it/s]


Model Was Saved ! Current Best Avg. Dice: 0.7398514677496517 Current Avg. Dice: 0.7398514677496517


Validate (476 / 10 Steps) (dice=0.73358): 100%|██████████| 17/17 [00:07<00:00,  2.26it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.7398514677496517 Current Avg. Dice: 0.7375317847027498


Validate (544 / 10 Steps) (dice=0.69298): 100%|██████████| 17/17 [00:07<00:00,  2.18it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.7398514677496517 Current Avg. Dice: 0.7320890742189744


Validate (680 / 10 Steps) (dice=0.71096): 100%|██████████| 17/17 [00:07<00:00,  2.20it/s]


Model Was Saved ! Current Best Avg. Dice: 0.7498616050271427 Current Avg. Dice: 0.7498616050271427


Validate (748 / 10 Steps) (dice=0.73103): 100%|██████████| 17/17 [00:07<00:00,  2.17it/s]


Model Was Saved ! Current Best Avg. Dice: 0.7528331980985754 Current Avg. Dice: 0.7528331980985754


Validate (884 / 10 Steps) (dice=0.70953): 100%|██████████| 17/17 [00:07<00:00,  2.20it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.7528331980985754 Current Avg. Dice: 0.7222202139742234


Validate (952 / 10 Steps) (dice=0.72275): 100%|██████████| 17/17 [00:07<00:00,  2.20it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.7528331980985754 Current Avg. Dice: 0.7495952564127305


Validate (1088 / 10 Steps) (dice=0.70748): 100%|██████████| 17/17 [00:07<00:00,  2.14it/s]


Model Was Saved ! Current Best Avg. Dice: 0.7541990245089811 Current Avg. Dice: 0.7541990245089811


Validate (1156 / 10 Steps) (dice=0.74744): 100%|██████████| 17/17 [00:07<00:00,  2.28it/s]


Model Was Saved ! Current Best Avg. Dice: 0.7850285452954909 Current Avg. Dice: 0.7850285452954909


Validate (1292 / 10 Steps) (dice=0.73583): 100%|██████████| 17/17 [00:08<00:00,  2.11it/s]


Model Was Saved ! Current Best Avg. Dice: 0.7860456424600938 Current Avg. Dice: 0.7860456424600938


Validate (1360 / 10 Steps) (dice=0.74918): 100%|██████████| 17/17 [00:07<00:00,  2.25it/s]


Model Was Saved ! Current Best Avg. Dice: 0.7914346176035264 Current Avg. Dice: 0.7914346176035264


Validate (1496 / 10 Steps) (dice=0.76545): 100%|██████████| 17/17 [00:07<00:00,  2.22it/s]


Model Was Saved ! Current Best Avg. Dice: 0.8065095298430499 Current Avg. Dice: 0.8065095298430499


Validate (1564 / 10 Steps) (dice=0.81011): 100%|██████████| 17/17 [00:08<00:00,  2.09it/s]


Model Was Saved ! Current Best Avg. Dice: 0.8485558558912838 Current Avg. Dice: 0.8485558558912838


Validate (1700 / 10 Steps) (dice=0.85169): 100%|██████████| 17/17 [00:07<00:00,  2.16it/s]


Model Was Saved ! Current Best Avg. Dice: 0.8764551702667686 Current Avg. Dice: 0.8764551702667686


Validate (1768 / 10 Steps) (dice=0.87703): 100%|██████████| 17/17 [00:08<00:00,  2.11it/s]


Model Was Saved ! Current Best Avg. Dice: 0.8861169219017029 Current Avg. Dice: 0.8861169219017029


Validate (1836 / 10 Steps) (dice=0.86430): 100%|██████████| 17/17 [00:07<00:00,  2.26it/s]


Model Was Saved ! Current Best Avg. Dice: 0.8946084695703843 Current Avg. Dice: 0.8946084695703843


Validate (1972 / 10 Steps) (dice=0.86273): 100%|██████████| 17/17 [00:07<00:00,  2.23it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.8946084695703843 Current Avg. Dice: 0.8733194154851577


Validate (2040 / 10 Steps) (dice=0.87858): 100%|██████████| 17/17 [00:07<00:00,  2.23it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.8946084695703843 Current Avg. Dice: 0.8935087638742784


Validate (2176 / 10 Steps) (dice=0.88583): 100%|██████████| 17/17 [00:07<00:00,  2.26it/s]


Model Was Saved ! Current Best Avg. Dice: 0.9051182831034941 Current Avg. Dice: 0.9051182831034941


Validate (2244 / 10 Steps) (dice=0.88404): 100%|██████████| 17/17 [00:07<00:00,  2.23it/s]


Model Was Saved ! Current Best Avg. Dice: 0.9083754911142237 Current Avg. Dice: 0.9083754911142237


Validate (2380 / 10 Steps) (dice=0.86812): 100%|██████████| 17/17 [00:07<00:00,  2.26it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.9083754911142237 Current Avg. Dice: 0.9000601698370541


Validate (2448 / 10 Steps) (dice=0.88979): 100%|██████████| 17/17 [00:07<00:00,  2.15it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.9083754911142237 Current Avg. Dice: 0.9062365258441252


Validate (2584 / 10 Steps) (dice=0.87010): 100%|██████████| 17/17 [00:07<00:00,  2.21it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.9083754911142237 Current Avg. Dice: 0.8968295384855831


Validate (2652 / 10 Steps) (dice=0.89153): 100%|██████████| 17/17 [00:07<00:00,  2.21it/s]


Model Was Saved ! Current Best Avg. Dice: 0.9196391245898079 Current Avg. Dice: 0.9196391245898079


Validate (2788 / 10 Steps) (dice=0.87224): 100%|██████████| 17/17 [00:07<00:00,  2.20it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.9196391245898079 Current Avg. Dice: 0.8980096543536467


Validate (2856 / 10 Steps) (dice=0.87269): 100%|██████████| 17/17 [00:07<00:00,  2.23it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.9196391245898079 Current Avg. Dice: 0.8957263231277466


Validate (2992 / 10 Steps) (dice=0.87325): 100%|██████████| 17/17 [00:07<00:00,  2.17it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.9196391245898079 Current Avg. Dice: 0.9025775825276094


Validate (3060 / 10 Steps) (dice=0.86530): 100%|██████████| 17/17 [00:07<00:00,  2.23it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.9196391245898079 Current Avg. Dice: 0.8958005659720477


Validate (3196 / 10 Steps) (dice=0.87347): 100%|██████████| 17/17 [00:07<00:00,  2.24it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.9196391245898079 Current Avg. Dice: 0.9006703986841089


Validate (3264 / 10 Steps) (dice=0.88178): 100%|██████████| 17/17 [00:07<00:00,  2.19it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.9196391245898079 Current Avg. Dice: 0.9034639106077307


Validate (3400 / 10 Steps) (dice=0.88192): 100%|██████████| 17/17 [00:07<00:00,  2.14it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.9196391245898079 Current Avg. Dice: 0.9031154127681956


Validate (3468 / 10 Steps) (dice=0.87451): 100%|██████████| 17/17 [00:07<00:00,  2.24it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.9196391245898079 Current Avg. Dice: 0.8940884576124304


Validate (3536 / 10 Steps) (dice=0.88624): 100%|██████████| 17/17 [00:07<00:00,  2.27it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.9196391245898079 Current Avg. Dice: 0.9114920286571279


Validate (3672 / 10 Steps) (dice=0.89132): 100%|██████████| 17/17 [00:07<00:00,  2.25it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.9196391245898079 Current Avg. Dice: 0.9168309849851272


Validate (3740 / 10 Steps) (dice=0.88433): 100%|██████████| 17/17 [00:07<00:00,  2.19it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.9196391245898079 Current Avg. Dice: 0.9141697638175067


Validate (3876 / 10 Steps) (dice=0.84111): 100%|██████████| 17/17 [00:08<00:00,  2.12it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.9196391245898079 Current Avg. Dice: 0.8691985712331884


Validate (3944 / 10 Steps) (dice=0.88655): 100%|██████████| 17/17 [00:07<00:00,  2.25it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.9196391245898079 Current Avg. Dice: 0.9069822044933543


Validate (4080 / 10 Steps) (dice=0.87090): 100%|██████████| 17/17 [00:07<00:00,  2.21it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.9196391245898079 Current Avg. Dice: 0.8878519149387584


Validate (4148 / 10 Steps) (dice=0.88578): 100%|██████████| 17/17 [00:07<00:00,  2.23it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.9196391245898079 Current Avg. Dice: 0.9085098505020142


Validate (4284 / 10 Steps) (dice=0.87179): 100%|██████████| 17/17 [00:07<00:00,  2.26it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.9196391245898079 Current Avg. Dice: 0.9032991458387936


Validate (4352 / 10 Steps) (dice=0.87611): 100%|██████████| 17/17 [00:08<00:00,  2.10it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.9196391245898079 Current Avg. Dice: 0.8996208029634812


Validate (4488 / 10 Steps) (dice=0.89145): 100%|██████████| 17/17 [00:07<00:00,  2.22it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.9196391245898079 Current Avg. Dice: 0.9060003196491915


Validate (4556 / 10 Steps) (dice=0.87173): 100%|██████████| 17/17 [00:08<00:00,  2.12it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.9196391245898079 Current Avg. Dice: 0.8974798286662382


Validate (4692 / 10 Steps) (dice=0.88429): 100%|██████████| 17/17 [00:07<00:00,  2.20it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.9196391245898079 Current Avg. Dice: 0.9111032170407912


Validate (4760 / 10 Steps) (dice=0.88137): 100%|██████████| 17/17 [00:07<00:00,  2.20it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.9196391245898079 Current Avg. Dice: 0.9037821994108313


Validate (4896 / 10 Steps) (dice=0.86072): 100%|██████████| 17/17 [00:07<00:00,  2.21it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.9196391245898079 Current Avg. Dice: 0.8987901596462026


Validate (4964 / 10 Steps) (dice=0.88308): 100%|██████████| 17/17 [00:07<00:00,  2.24it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.9196391245898079 Current Avg. Dice: 0.9078511490541346


Validate (5100 / 10 Steps) (dice=0.88447): 100%|██████████| 17/17 [00:07<00:00,  2.15it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.9196391245898079 Current Avg. Dice: 0.9075794290093815


Validate (5168 / 10 Steps) (dice=0.89358): 100%|██████████| 17/17 [00:07<00:00,  2.19it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.9196391245898079 Current Avg. Dice: 0.9132847505457261


Validate (5236 / 10 Steps) (dice=0.86864): 100%|██████████| 17/17 [00:07<00:00,  2.26it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.9196391245898079 Current Avg. Dice: 0.8886881540803349


Validate (5372 / 10 Steps) (dice=0.87063): 100%|██████████| 17/17 [00:07<00:00,  2.20it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.9196391245898079 Current Avg. Dice: 0.8873530521112329


Validate (5440 / 10 Steps) (dice=0.88764): 100%|██████████| 17/17 [00:07<00:00,  2.15it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.9196391245898079 Current Avg. Dice: 0.9135723254259895


Validate (5576 / 10 Steps) (dice=0.89362): 100%|██████████| 17/17 [00:07<00:00,  2.19it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.9196391245898079 Current Avg. Dice: 0.9177607087528005


Validate (5644 / 10 Steps) (dice=0.88907): 100%|██████████| 17/17 [00:07<00:00,  2.13it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.9196391245898079 Current Avg. Dice: 0.9052111786954543


Validate (5780 / 10 Steps) (dice=0.87798): 100%|██████████| 17/17 [00:07<00:00,  2.15it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.9196391245898079 Current Avg. Dice: 0.897168755531311


Validate (5848 / 10 Steps) (dice=0.89208): 100%|██████████| 17/17 [00:07<00:00,  2.14it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.9196391245898079 Current Avg. Dice: 0.9031723211793339


Validate (5984 / 10 Steps) (dice=0.88516): 100%|██████████| 17/17 [00:08<00:00,  2.12it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.9196391245898079 Current Avg. Dice: 0.901697625132168


Validate (6052 / 10 Steps) (dice=0.87156): 100%|██████████| 17/17 [00:07<00:00,  2.22it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.9196391245898079 Current Avg. Dice: 0.8967830047887915


Validate (6188 / 10 Steps) (dice=0.89004): 100%|██████████| 17/17 [00:08<00:00,  2.08it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.9196391245898079 Current Avg. Dice: 0.9104573621469385


Validate (6256 / 10 Steps) (dice=0.86856): 100%|██████████| 17/17 [00:07<00:00,  2.20it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.9196391245898079 Current Avg. Dice: 0.8865422361037311


Validate (6392 / 10 Steps) (dice=0.88819): 100%|██████████| 17/17 [00:07<00:00,  2.20it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.9196391245898079 Current Avg. Dice: 0.9023771040579852


Validate (6460 / 10 Steps) (dice=0.87174): 100%|██████████| 17/17 [00:07<00:00,  2.22it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.9196391245898079 Current Avg. Dice: 0.8812424260027268


Validate (6596 / 10 Steps) (dice=0.88987): 100%|██████████| 17/17 [00:07<00:00,  2.23it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.9196391245898079 Current Avg. Dice: 0.9121836388812345


Validate (6664 / 10 Steps) (dice=0.89641): 100%|██████████| 17/17 [00:07<00:00,  2.24it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.9196391245898079 Current Avg. Dice: 0.9130354804151198


Validate (6800 / 10 Steps) (dice=0.89138): 100%|██████████| 17/17 [00:07<00:00,  2.16it/s]


Model Was Saved ! Current Best Avg. Dice: 0.9223889533211204 Current Avg. Dice: 0.9223889533211204


Validate (6868 / 10 Steps) (dice=0.88455): 100%|██████████| 17/17 [00:07<00:00,  2.20it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.9223889533211204 Current Avg. Dice: 0.8925681289504556


Validate (6936 / 10 Steps) (dice=0.88513): 100%|██████████| 17/17 [00:07<00:00,  2.24it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.9223889533211204 Current Avg. Dice: 0.9070143489276662


Validate (7072 / 10 Steps) (dice=0.86536): 100%|██████████| 17/17 [00:07<00:00,  2.24it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.9223889533211204 Current Avg. Dice: 0.8848199493744794


Validate (7140 / 10 Steps) (dice=0.88218): 100%|██████████| 17/17 [00:07<00:00,  2.22it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.9223889533211204 Current Avg. Dice: 0.9133888973909265


Validate (7276 / 10 Steps) (dice=0.88958): 100%|██████████| 17/17 [00:07<00:00,  2.16it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.9223889533211204 Current Avg. Dice: 0.9056829179034513


Validate (7344 / 10 Steps) (dice=0.89404): 100%|██████████| 17/17 [00:07<00:00,  2.14it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.9223889533211204 Current Avg. Dice: 0.9147317269269157


Validate (7480 / 10 Steps) (dice=0.87597): 100%|██████████| 17/17 [00:07<00:00,  2.21it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.9223889533211204 Current Avg. Dice: 0.902138047358569


Validate (7548 / 10 Steps) (dice=0.88991): 100%|██████████| 17/17 [00:07<00:00,  2.21it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.9223889533211204 Current Avg. Dice: 0.9055923223495483


Validate (7684 / 10 Steps) (dice=0.88669): 100%|██████████| 17/17 [00:07<00:00,  2.23it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.9223889533211204 Current Avg. Dice: 0.9115563490811516


Validate (7752 / 10 Steps) (dice=0.88903): 100%|██████████| 17/17 [00:08<00:00,  2.12it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.9223889533211204 Current Avg. Dice: 0.9098890528959387


Validate (7888 / 10 Steps) (dice=0.88619): 100%|██████████| 17/17 [00:07<00:00,  2.16it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.9223889533211204 Current Avg. Dice: 0.8973058356958277


Validate (7956 / 10 Steps) (dice=0.89325): 100%|██████████| 17/17 [00:07<00:00,  2.25it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.9223889533211204 Current Avg. Dice: 0.9054997668546789


Validate (8092 / 10 Steps) (dice=0.89074): 100%|██████████| 17/17 [00:07<00:00,  2.21it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.9223889533211204 Current Avg. Dice: 0.9080874604337356


Validate (8160 / 10 Steps) (dice=0.90573): 100%|██████████| 17/17 [00:07<00:00,  2.22it/s]


Model Was Saved ! Current Best Avg. Dice: 0.925990385167739 Current Avg. Dice: 0.925990385167739


Validate (8296 / 10 Steps) (dice=0.90206): 100%|██████████| 17/17 [00:07<00:00,  2.25it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.925990385167739 Current Avg. Dice: 0.9107935744173387


Validate (8364 / 10 Steps) (dice=0.89282): 100%|██████████| 17/17 [00:07<00:00,  2.16it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.925990385167739 Current Avg. Dice: 0.9104833743151497


Validate (8500 / 10 Steps) (dice=0.90296): 100%|██████████| 17/17 [00:07<00:00,  2.14it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.925990385167739 Current Avg. Dice: 0.9238737646271201


Validate (8568 / 10 Steps) (dice=0.90194): 100%|██████████| 17/17 [00:07<00:00,  2.19it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.925990385167739 Current Avg. Dice: 0.9221123316708733


Validate (8636 / 10 Steps) (dice=0.91527): 100%|██████████| 17/17 [00:07<00:00,  2.26it/s]


Model Was Saved ! Current Best Avg. Dice: 0.9379014618256513 Current Avg. Dice: 0.9379014618256513


Validate (8772 / 10 Steps) (dice=0.89499): 100%|██████████| 17/17 [00:07<00:00,  2.25it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.9379014618256513 Current Avg. Dice: 0.9160064528970158


Validate (8840 / 10 Steps) (dice=0.90779): 100%|██████████| 17/17 [00:07<00:00,  2.28it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.9379014618256513 Current Avg. Dice: 0.9124572031638202


Validate (8976 / 10 Steps) (dice=0.89080): 100%|██████████| 17/17 [00:07<00:00,  2.23it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.9379014618256513 Current Avg. Dice: 0.9087486232028288


Validate (9044 / 10 Steps) (dice=0.91054): 100%|██████████| 17/17 [00:07<00:00,  2.15it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.9379014618256513 Current Avg. Dice: 0.9208257759318632


Validate (9180 / 10 Steps) (dice=0.90509): 100%|██████████| 17/17 [00:07<00:00,  2.16it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.9379014618256513 Current Avg. Dice: 0.9154434344347786


Validate (9248 / 10 Steps) (dice=0.90009): 100%|██████████| 17/17 [00:08<00:00,  2.11it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.9379014618256513 Current Avg. Dice: 0.921119353350471


Validate (9384 / 10 Steps) (dice=0.87852): 100%|██████████| 17/17 [00:07<00:00,  2.20it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.9379014618256513 Current Avg. Dice: 0.9011624210021075


Validate (9452 / 10 Steps) (dice=0.90721): 100%|██████████| 17/17 [00:07<00:00,  2.20it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.9379014618256513 Current Avg. Dice: 0.9244220677544089


Validate (9588 / 10 Steps) (dice=0.90126): 100%|██████████| 17/17 [00:07<00:00,  2.22it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.9379014618256513 Current Avg. Dice: 0.9141853451728821


Validate (9656 / 10 Steps) (dice=0.90219): 100%|██████████| 17/17 [00:08<00:00,  2.05it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.9379014618256513 Current Avg. Dice: 0.9258808388429529


Validate (9792 / 10 Steps) (dice=0.88552): 100%|██████████| 17/17 [00:07<00:00,  2.13it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.9379014618256513 Current Avg. Dice: 0.90610872296726


Validate (9860 / 10 Steps) (dice=0.89056): 100%|██████████| 17/17 [00:08<00:00,  2.12it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.9379014618256513 Current Avg. Dice: 0.9124549557180965


Validate (9996 / 10 Steps) (dice=0.90991): 100%|██████████| 17/17 [00:07<00:00,  2.18it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.9379014618256513 Current Avg. Dice: 0.9274025138686685


Training (10063 / 10000 Steps) (loss=0.44547): 100%|██████████| 68/68 [00:37<00:00,  1.80it/s]


Fold 1/5 completed, best_metric: 0.9379 at iteration: 8700

Fold 2/5


Loading dataset: 100%|██████████| 6/6 [00:04<00:00,  1.30it/s]


Loading Weights from the Path /content/drive/MyDrive/template/our-git-project/Pretrained_Results_final_v2/best_model.pt
Pretrained Weights Succesfully Loaded !


Validate (68 / 10 Steps) (dice=0.48190): 100%|██████████| 17/17 [00:07<00:00,  2.15it/s]


Model Was Saved ! Current Best Avg. Dice: 0.4739447411368875 Current Avg. Dice: 0.4739447411368875


Validate (136 / 10 Steps) (dice=0.55125): 100%|██████████| 17/17 [00:07<00:00,  2.27it/s]


Model Was Saved ! Current Best Avg. Dice: 0.5478923566201154 Current Avg. Dice: 0.5478923566201154


Validate (272 / 10 Steps) (dice=0.57524): 100%|██████████| 17/17 [00:07<00:00,  2.13it/s]


Model Was Saved ! Current Best Avg. Dice: 0.5590901444940006 Current Avg. Dice: 0.5590901444940006


Validate (340 / 10 Steps) (dice=0.60498): 100%|██████████| 17/17 [00:07<00:00,  2.21it/s]


Model Was Saved ! Current Best Avg. Dice: 0.600716909941505 Current Avg. Dice: 0.600716909941505


Validate (476 / 10 Steps) (dice=0.61109): 100%|██████████| 17/17 [00:07<00:00,  2.18it/s]


Model Was Saved ! Current Best Avg. Dice: 0.6153262222514433 Current Avg. Dice: 0.6153262222514433


Validate (544 / 10 Steps) (dice=0.63392): 100%|██████████| 17/17 [00:07<00:00,  2.16it/s]


Model Was Saved ! Current Best Avg. Dice: 0.6412801917861489 Current Avg. Dice: 0.6412801917861489


Validate (680 / 10 Steps) (dice=0.67930): 100%|██████████| 17/17 [00:07<00:00,  2.20it/s]


Model Was Saved ! Current Best Avg. Dice: 0.6591856584829443 Current Avg. Dice: 0.6591856584829443


Validate (748 / 10 Steps) (dice=0.68857): 100%|██████████| 17/17 [00:07<00:00,  2.17it/s]


Model Was Saved ! Current Best Avg. Dice: 0.6728330710354973 Current Avg. Dice: 0.6728330710354973


Validate (884 / 10 Steps) (dice=0.69657): 100%|██████████| 17/17 [00:07<00:00,  2.17it/s]


Model Was Saved ! Current Best Avg. Dice: 0.6973866750212276 Current Avg. Dice: 0.6973866750212276


Validate (952 / 10 Steps) (dice=0.68343): 100%|██████████| 17/17 [00:07<00:00,  2.14it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.6973866750212276 Current Avg. Dice: 0.6623974652851329


Validate (1088 / 10 Steps) (dice=0.71754): 100%|██████████| 17/17 [00:08<00:00,  2.06it/s]


Model Was Saved ! Current Best Avg. Dice: 0.7135517772506265 Current Avg. Dice: 0.7135517772506265


Validate (1156 / 10 Steps) (dice=0.70908): 100%|██████████| 17/17 [00:07<00:00,  2.21it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.7135517772506265 Current Avg. Dice: 0.6947663987384123


Validate (1292 / 10 Steps) (dice=0.70938): 100%|██████████| 17/17 [00:07<00:00,  2.25it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.7135517772506265 Current Avg. Dice: 0.708272355444291


Validate (1360 / 10 Steps) (dice=0.75402): 100%|██████████| 17/17 [00:07<00:00,  2.23it/s]


Model Was Saved ! Current Best Avg. Dice: 0.7492526068406946 Current Avg. Dice: 0.7492526068406946


Validate (1496 / 10 Steps) (dice=0.75658): 100%|██████████| 17/17 [00:07<00:00,  2.15it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.7492526068406946 Current Avg. Dice: 0.733538988758536


Validate (1564 / 10 Steps) (dice=0.75565): 100%|██████████| 17/17 [00:08<00:00,  2.11it/s]


Model Was Saved ! Current Best Avg. Dice: 0.7825174366726595 Current Avg. Dice: 0.7825174366726595


Validate (1700 / 10 Steps) (dice=0.76008): 100%|██████████| 17/17 [00:07<00:00,  2.20it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.7825174366726595 Current Avg. Dice: 0.7682152180110707


Validate (1768 / 10 Steps) (dice=0.79967): 100%|██████████| 17/17 [00:07<00:00,  2.22it/s]


Model Was Saved ! Current Best Avg. Dice: 0.7861968243823332 Current Avg. Dice: 0.7861968243823332


Validate (1836 / 10 Steps) (dice=0.78744): 100%|██████████| 17/17 [00:07<00:00,  2.22it/s]


Model Was Saved ! Current Best Avg. Dice: 0.7994432238971486 Current Avg. Dice: 0.7994432238971486


Validate (1972 / 10 Steps) (dice=0.81398): 100%|██████████| 17/17 [00:07<00:00,  2.16it/s]


Model Was Saved ! Current Best Avg. Dice: 0.8365541345932904 Current Avg. Dice: 0.8365541345932904


Validate (2040 / 10 Steps) (dice=0.82466): 100%|██████████| 17/17 [00:07<00:00,  2.21it/s]


Model Was Saved ! Current Best Avg. Dice: 0.8462932705879211 Current Avg. Dice: 0.8462932705879211


Validate (2176 / 10 Steps) (dice=0.83468): 100%|██████████| 17/17 [00:07<00:00,  2.15it/s]


Model Was Saved ! Current Best Avg. Dice: 0.8463352603070876 Current Avg. Dice: 0.8463352603070876


Validate (2244 / 10 Steps) (dice=0.81908): 100%|██████████| 17/17 [00:07<00:00,  2.14it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.8463352603070876 Current Avg. Dice: 0.828280557604397


Validate (2380 / 10 Steps) (dice=0.83524): 100%|██████████| 17/17 [00:07<00:00,  2.18it/s]


Model Was Saved ! Current Best Avg. Dice: 0.8539834162768196 Current Avg. Dice: 0.8539834162768196


Validate (2448 / 10 Steps) (dice=0.82988): 100%|██████████| 17/17 [00:08<00:00,  2.06it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.8539834162768196 Current Avg. Dice: 0.8354523672777063


Validate (2584 / 10 Steps) (dice=0.84711): 100%|██████████| 17/17 [00:07<00:00,  2.21it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.8539834162768196 Current Avg. Dice: 0.8491329761112437


Validate (2652 / 10 Steps) (dice=0.82762): 100%|██████████| 17/17 [00:07<00:00,  2.19it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.8539834162768196 Current Avg. Dice: 0.8403780460357666


Validate (2788 / 10 Steps) (dice=0.83636): 100%|██████████| 17/17 [00:07<00:00,  2.15it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.8539834162768196 Current Avg. Dice: 0.846065153093899


Validate (2856 / 10 Steps) (dice=0.83564): 100%|██████████| 17/17 [00:08<00:00,  2.05it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.8539834162768196 Current Avg. Dice: 0.8449603213983423


Validate (2992 / 10 Steps) (dice=0.87437): 100%|██████████| 17/17 [00:07<00:00,  2.22it/s]


Model Was Saved ! Current Best Avg. Dice: 0.8778234965660993 Current Avg. Dice: 0.8778234965660993


Validate (3060 / 10 Steps) (dice=0.85618): 100%|██████████| 17/17 [00:07<00:00,  2.19it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.8778234965660993 Current Avg. Dice: 0.847628092064577


Validate (3196 / 10 Steps) (dice=0.84143): 100%|██████████| 17/17 [00:07<00:00,  2.20it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.8778234965660993 Current Avg. Dice: 0.8522519854938283


Validate (3264 / 10 Steps) (dice=0.84253): 100%|██████████| 17/17 [00:07<00:00,  2.19it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.8778234965660993 Current Avg. Dice: 0.866355696145226


Validate (3400 / 10 Steps) (dice=0.83555): 100%|██████████| 17/17 [00:07<00:00,  2.13it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.8778234965660993 Current Avg. Dice: 0.8406844279345345


Validate (3468 / 10 Steps) (dice=0.85535): 100%|██████████| 17/17 [00:08<00:00,  2.09it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.8778234965660993 Current Avg. Dice: 0.865384487544789


Validate (3536 / 10 Steps) (dice=0.81346): 100%|██████████| 17/17 [00:07<00:00,  2.23it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.8778234965660993 Current Avg. Dice: 0.8219119906425476


Validate (3672 / 10 Steps) (dice=0.85195): 100%|██████████| 17/17 [00:07<00:00,  2.18it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.8778234965660993 Current Avg. Dice: 0.865475735243629


Validate (3740 / 10 Steps) (dice=0.84581): 100%|██████████| 17/17 [00:07<00:00,  2.21it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.8778234965660993 Current Avg. Dice: 0.8703540072721594


Validate (3876 / 10 Steps) (dice=0.82755): 100%|██████████| 17/17 [00:07<00:00,  2.14it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.8778234965660993 Current Avg. Dice: 0.8422233918133903


Validate (3944 / 10 Steps) (dice=0.85189): 100%|██████████| 17/17 [00:07<00:00,  2.14it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.8778234965660993 Current Avg. Dice: 0.8713587171891156


Validate (4080 / 10 Steps) (dice=0.84008): 100%|██████████| 17/17 [00:07<00:00,  2.24it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.8778234965660993 Current Avg. Dice: 0.8403235428473529


Validate (4148 / 10 Steps) (dice=0.85212): 100%|██████████| 17/17 [00:08<00:00,  2.05it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.8778234965660993 Current Avg. Dice: 0.8359739710302914


Validate (4284 / 10 Steps) (dice=0.84716): 100%|██████████| 17/17 [00:07<00:00,  2.13it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.8778234965660993 Current Avg. Dice: 0.8651183948797339


Validate (4352 / 10 Steps) (dice=0.85826): 100%|██████████| 17/17 [00:08<00:00,  2.10it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.8778234965660993 Current Avg. Dice: 0.8701919843168819


Validate (4488 / 10 Steps) (dice=0.85976): 100%|██████████| 17/17 [00:07<00:00,  2.17it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.8778234965660993 Current Avg. Dice: 0.8679998481974882


Validate (4556 / 10 Steps) (dice=0.85174): 100%|██████████| 17/17 [00:07<00:00,  2.17it/s]


Model Was Saved ! Current Best Avg. Dice: 0.8800405859947205 Current Avg. Dice: 0.8800405859947205


Validate (4692 / 10 Steps) (dice=0.83741): 100%|██████████| 17/17 [00:07<00:00,  2.13it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.8800405859947205 Current Avg. Dice: 0.8424043129472172


Validate (4760 / 10 Steps) (dice=0.82238): 100%|██████████| 17/17 [00:07<00:00,  2.14it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.8800405859947205 Current Avg. Dice: 0.8383457660675049


Validate (4896 / 10 Steps) (dice=0.83794): 100%|██████████| 17/17 [00:07<00:00,  2.24it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.8800405859947205 Current Avg. Dice: 0.8633754183264339


Validate (4964 / 10 Steps) (dice=0.84469): 100%|██████████| 17/17 [00:07<00:00,  2.16it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.8800405859947205 Current Avg. Dice: 0.8705050033681533


Validate (5100 / 10 Steps) (dice=0.86388): 100%|██████████| 17/17 [00:08<00:00,  2.11it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.8800405859947205 Current Avg. Dice: 0.8732472237418679


Validate (5168 / 10 Steps) (dice=0.87034): 100%|██████████| 17/17 [00:07<00:00,  2.19it/s]


Model Was Saved ! Current Best Avg. Dice: 0.883830585900475 Current Avg. Dice: 0.883830585900475


Validate (5236 / 10 Steps) (dice=0.84294): 100%|██████████| 17/17 [00:07<00:00,  2.17it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.883830585900475 Current Avg. Dice: 0.8711498870569117


Validate (5372 / 10 Steps) (dice=0.84303): 100%|██████████| 17/17 [00:08<00:00,  2.11it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.883830585900475 Current Avg. Dice: 0.8489649646422442


Validate (5440 / 10 Steps) (dice=0.85616): 100%|██████████| 17/17 [00:07<00:00,  2.21it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.883830585900475 Current Avg. Dice: 0.8507748982485603


Validate (5576 / 10 Steps) (dice=0.85701): 100%|██████████| 17/17 [00:07<00:00,  2.17it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.883830585900475 Current Avg. Dice: 0.864527653245365


Validate (5644 / 10 Steps) (dice=0.84397): 100%|██████████| 17/17 [00:08<00:00,  2.09it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.883830585900475 Current Avg. Dice: 0.873377354706035


Validate (5780 / 10 Steps) (dice=0.84749): 100%|██████████| 17/17 [00:07<00:00,  2.20it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.883830585900475 Current Avg. Dice: 0.8695727586746216


Validate (5848 / 10 Steps) (dice=0.85516): 100%|██████████| 17/17 [00:07<00:00,  2.22it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.883830585900475 Current Avg. Dice: 0.8633823605144725


Validate (5984 / 10 Steps) (dice=0.84570): 100%|██████████| 17/17 [00:08<00:00,  2.12it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.883830585900475 Current Avg. Dice: 0.8627047433572657


Validate (6052 / 10 Steps) (dice=0.85498): 100%|██████████| 17/17 [00:07<00:00,  2.19it/s]


Model Was Saved ! Current Best Avg. Dice: 0.8924739045255324 Current Avg. Dice: 0.8924739045255324


Validate (6188 / 10 Steps) (dice=0.86370): 100%|██████████| 17/17 [00:07<00:00,  2.20it/s]


Model Was Saved ! Current Best Avg. Dice: 0.8939325073186088 Current Avg. Dice: 0.8939325073186088


Validate (6256 / 10 Steps) (dice=0.85543): 100%|██████████| 17/17 [00:08<00:00,  2.11it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.8939325073186088 Current Avg. Dice: 0.8742993018206429


Validate (6392 / 10 Steps) (dice=0.85026): 100%|██████████| 17/17 [00:07<00:00,  2.24it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.8939325073186088 Current Avg. Dice: 0.8770851983743555


Validate (6460 / 10 Steps) (dice=0.86308): 100%|██████████| 17/17 [00:07<00:00,  2.24it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.8939325073186088 Current Avg. Dice: 0.8916365889941945


Validate (6596 / 10 Steps) (dice=0.85857): 100%|██████████| 17/17 [00:08<00:00,  2.10it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.8939325073186088 Current Avg. Dice: 0.8835060000419617


Validate (6664 / 10 Steps) (dice=0.85603): 100%|██████████| 17/17 [00:07<00:00,  2.18it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.8939325073186088 Current Avg. Dice: 0.8769653123967788


Validate (6800 / 10 Steps) (dice=0.85157): 100%|██████████| 17/17 [00:07<00:00,  2.20it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.8939325073186088 Current Avg. Dice: 0.8768406194799087


Validate (6868 / 10 Steps) (dice=0.86226): 100%|██████████| 17/17 [00:07<00:00,  2.16it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.8939325073186088 Current Avg. Dice: 0.886095222304849


Validate (6936 / 10 Steps) (dice=0.84881): 100%|██████████| 17/17 [00:07<00:00,  2.23it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.8939325073186088 Current Avg. Dice: 0.8730599634787616


Validate (7072 / 10 Steps) (dice=0.86411): 100%|██████████| 17/17 [00:08<00:00,  2.10it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.8939325073186088 Current Avg. Dice: 0.8933205359122333


Validate (7140 / 10 Steps) (dice=0.85266): 100%|██████████| 17/17 [00:07<00:00,  2.22it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.8939325073186088 Current Avg. Dice: 0.8640729855088627


Validate (7276 / 10 Steps) (dice=0.85793): 100%|██████████| 17/17 [00:08<00:00,  2.11it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.8939325073186088 Current Avg. Dice: 0.880295504541958


Validate (7344 / 10 Steps) (dice=0.85393): 100%|██████████| 17/17 [00:07<00:00,  2.14it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.8939325073186088 Current Avg. Dice: 0.8626462606822743


Validate (7480 / 10 Steps) (dice=0.87792): 100%|██████████| 17/17 [00:07<00:00,  2.22it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.8939325073186088 Current Avg. Dice: 0.8932209400569692


Validate (7548 / 10 Steps) (dice=0.86749): 100%|██████████| 17/17 [00:07<00:00,  2.19it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.8939325073186088 Current Avg. Dice: 0.8677658789298114


Validate (7684 / 10 Steps) (dice=0.87141): 100%|██████████| 17/17 [00:07<00:00,  2.21it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.8939325073186088 Current Avg. Dice: 0.8792790770530701


Validate (7752 / 10 Steps) (dice=0.87228): 100%|██████████| 17/17 [00:08<00:00,  2.12it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.8939325073186088 Current Avg. Dice: 0.8904007988817552


Validate (7888 / 10 Steps) (dice=0.87510): 100%|██████████| 17/17 [00:07<00:00,  2.14it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.8939325073186088 Current Avg. Dice: 0.8790466680246241


Validate (7956 / 10 Steps) (dice=0.88167): 100%|██████████| 17/17 [00:07<00:00,  2.17it/s]


Model Was Saved ! Current Best Avg. Dice: 0.9018226265907288 Current Avg. Dice: 0.9018226265907288


Validate (8092 / 10 Steps) (dice=0.86633): 100%|██████████| 17/17 [00:07<00:00,  2.21it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.9018226265907288 Current Avg. Dice: 0.8890850999776054


Validate (8160 / 10 Steps) (dice=0.86248): 100%|██████████| 17/17 [00:07<00:00,  2.18it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.9018226265907288 Current Avg. Dice: 0.879701200653525


Validate (8296 / 10 Steps) (dice=0.87629): 100%|██████████| 17/17 [00:07<00:00,  2.20it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.9018226265907288 Current Avg. Dice: 0.893410184804131


Validate (8364 / 10 Steps) (dice=0.85096): 100%|██████████| 17/17 [00:07<00:00,  2.19it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.9018226265907288 Current Avg. Dice: 0.8667693383553449


Validate (8500 / 10 Steps) (dice=0.86139): 100%|██████████| 17/17 [00:07<00:00,  2.21it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.9018226265907288 Current Avg. Dice: 0.8876638307290918


Validate (8568 / 10 Steps) (dice=0.87268): 100%|██████████| 17/17 [00:07<00:00,  2.21it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.9018226265907288 Current Avg. Dice: 0.8776369410402635


Validate (8636 / 10 Steps) (dice=0.86242): 100%|██████████| 17/17 [00:07<00:00,  2.24it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.9018226265907288 Current Avg. Dice: 0.8893437525805306


Validate (8772 / 10 Steps) (dice=0.88034): 100%|██████████| 17/17 [00:07<00:00,  2.21it/s]


Model Was Saved ! Current Best Avg. Dice: 0.9121307274874519 Current Avg. Dice: 0.9121307274874519


Validate (8840 / 10 Steps) (dice=0.90014): 100%|██████████| 17/17 [00:07<00:00,  2.18it/s]


Model Was Saved ! Current Best Avg. Dice: 0.9169705054339241 Current Avg. Dice: 0.9169705054339241


Validate (8976 / 10 Steps) (dice=0.86524): 100%|██████████| 17/17 [00:07<00:00,  2.17it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.9169705054339241 Current Avg. Dice: 0.8735966542187859


Validate (9044 / 10 Steps) (dice=0.86621): 100%|██████████| 17/17 [00:07<00:00,  2.18it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.9169705054339241 Current Avg. Dice: 0.897134377675898


Validate (9180 / 10 Steps) (dice=0.87944): 100%|██████████| 17/17 [00:08<00:00,  2.11it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.9169705054339241 Current Avg. Dice: 0.8913016704952016


Validate (9248 / 10 Steps) (dice=0.87203): 100%|██████████| 17/17 [00:07<00:00,  2.13it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.9169705054339241 Current Avg. Dice: 0.9011072656687569


Validate (9384 / 10 Steps) (dice=0.86958): 100%|██████████| 17/17 [00:08<00:00,  2.12it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.9169705054339241 Current Avg. Dice: 0.888053347082699


Validate (9452 / 10 Steps) (dice=0.86991): 100%|██████████| 17/17 [00:07<00:00,  2.18it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.9169705054339241 Current Avg. Dice: 0.900644014863407


Validate (9588 / 10 Steps) (dice=0.84518): 100%|██████████| 17/17 [00:07<00:00,  2.14it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.9169705054339241 Current Avg. Dice: 0.8381770989474129


Validate (9656 / 10 Steps) (dice=0.86641): 100%|██████████| 17/17 [00:07<00:00,  2.16it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.9169705054339241 Current Avg. Dice: 0.8782226267982932


Validate (9792 / 10 Steps) (dice=0.87147): 100%|██████████| 17/17 [00:07<00:00,  2.19it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.9169705054339241 Current Avg. Dice: 0.8970463451217202


Validate (9860 / 10 Steps) (dice=0.88617): 100%|██████████| 17/17 [00:07<00:00,  2.18it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.9169705054339241 Current Avg. Dice: 0.8785307793056264


Validate (9996 / 10 Steps) (dice=0.85841): 100%|██████████| 17/17 [00:08<00:00,  2.07it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.9169705054339241 Current Avg. Dice: 0.8761069704504574


Training (10063 / 10000 Steps) (loss=0.56648): 100%|██████████| 68/68 [00:38<00:00,  1.77it/s]


Fold 2/5 completed, best_metric: 0.9170 at iteration: 8900

Fold 3/5


Loading dataset: 100%|██████████| 6/6 [00:04<00:00,  1.31it/s]


Loading Weights from the Path /content/drive/MyDrive/template/our-git-project/Pretrained_Results_final_v2/best_model.pt
Pretrained Weights Succesfully Loaded !


Validate (68 / 10 Steps) (dice=0.48382): 100%|██████████| 17/17 [00:07<00:00,  2.18it/s]


Model Was Saved ! Current Best Avg. Dice: 0.4790608392042272 Current Avg. Dice: 0.4790608392042272


Validate (136 / 10 Steps) (dice=0.56871): 100%|██████████| 17/17 [00:07<00:00,  2.19it/s]


Model Was Saved ! Current Best Avg. Dice: 0.5684224752818837 Current Avg. Dice: 0.5684224752818837


Validate (272 / 10 Steps) (dice=0.57837): 100%|██████████| 17/17 [00:07<00:00,  2.16it/s]


Model Was Saved ! Current Best Avg. Dice: 0.5889573482906117 Current Avg. Dice: 0.5889573482906117


Validate (340 / 10 Steps) (dice=0.62378): 100%|██████████| 17/17 [00:08<00:00,  2.12it/s]


Model Was Saved ! Current Best Avg. Dice: 0.6251286688972922 Current Avg. Dice: 0.6251286688972922


Validate (476 / 10 Steps) (dice=0.60045): 100%|██████████| 17/17 [00:07<00:00,  2.17it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.6251286688972922 Current Avg. Dice: 0.6087056749007281


Validate (544 / 10 Steps) (dice=0.63577): 100%|██████████| 17/17 [00:08<00:00,  2.12it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.6251286688972922 Current Avg. Dice: 0.6072079258806565


Validate (680 / 10 Steps) (dice=0.59548): 100%|██████████| 17/17 [00:07<00:00,  2.15it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.6251286688972922 Current Avg. Dice: 0.5792845873271718


Validate (748 / 10 Steps) (dice=0.60876): 100%|██████████| 17/17 [00:07<00:00,  2.18it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.6251286688972922 Current Avg. Dice: 0.5977616625673631


Validate (884 / 10 Steps) (dice=0.60991): 100%|██████████| 17/17 [00:08<00:00,  2.09it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.6251286688972922 Current Avg. Dice: 0.6012114005930284


Validate (952 / 10 Steps) (dice=0.60535): 100%|██████████| 17/17 [00:07<00:00,  2.21it/s]


Model Was Saved ! Current Best Avg. Dice: 0.6279350098441628 Current Avg. Dice: 0.6279350098441628


Validate (1088 / 10 Steps) (dice=0.62737): 100%|██████████| 17/17 [00:08<00:00,  2.08it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.6279350098441628 Current Avg. Dice: 0.6164295778555029


Validate (1156 / 10 Steps) (dice=0.63271): 100%|██████████| 17/17 [00:07<00:00,  2.18it/s]


Model Was Saved ! Current Best Avg. Dice: 0.6291964860523448 Current Avg. Dice: 0.6291964860523448


Validate (1292 / 10 Steps) (dice=0.61567): 100%|██████████| 17/17 [00:08<00:00,  2.11it/s]


Model Was Saved ! Current Best Avg. Dice: 0.6296389348366681 Current Avg. Dice: 0.6296389348366681


Validate (1360 / 10 Steps) (dice=0.59448): 100%|██████████| 17/17 [00:07<00:00,  2.16it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.6296389348366681 Current Avg. Dice: 0.5639366086791543


Validate (1496 / 10 Steps) (dice=0.60625): 100%|██████████| 17/17 [00:07<00:00,  2.17it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.6296389348366681 Current Avg. Dice: 0.5885576009750366


Validate (1564 / 10 Steps) (dice=0.61211): 100%|██████████| 17/17 [00:07<00:00,  2.13it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.6296389348366681 Current Avg. Dice: 0.6117326757487129


Validate (1700 / 10 Steps) (dice=0.64839): 100%|██████████| 17/17 [00:07<00:00,  2.16it/s]


Model Was Saved ! Current Best Avg. Dice: 0.6604314481510836 Current Avg. Dice: 0.6604314481510836


Validate (1768 / 10 Steps) (dice=0.60802): 100%|██████████| 17/17 [00:08<00:00,  2.10it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.6604314481510836 Current Avg. Dice: 0.6090643265668083


Validate (1836 / 10 Steps) (dice=0.64662): 100%|██████████| 17/17 [00:07<00:00,  2.20it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.6604314481510836 Current Avg. Dice: 0.6366972642786363


Validate (1972 / 10 Steps) (dice=0.66825): 100%|██████████| 17/17 [00:07<00:00,  2.15it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.6604314481510836 Current Avg. Dice: 0.6545214442645803


Validate (2040 / 10 Steps) (dice=0.68496): 100%|██████████| 17/17 [00:08<00:00,  2.08it/s]


Model Was Saved ! Current Best Avg. Dice: 0.6953322081004872 Current Avg. Dice: 0.6953322081004872


Validate (2176 / 10 Steps) (dice=0.66255): 100%|██████████| 17/17 [00:07<00:00,  2.15it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.6953322081004872 Current Avg. Dice: 0.6562242087195901


Validate (2244 / 10 Steps) (dice=0.68396): 100%|██████████| 17/17 [00:08<00:00,  2.08it/s]


Model Was Saved ! Current Best Avg. Dice: 0.7335874157793382 Current Avg. Dice: 0.7335874157793382


Validate (2380 / 10 Steps) (dice=0.62511): 100%|██████████| 17/17 [00:08<00:00,  2.06it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.7335874157793382 Current Avg. Dice: 0.6157544430564431


Validate (2448 / 10 Steps) (dice=0.68142): 100%|██████████| 17/17 [00:07<00:00,  2.16it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.7335874157793382 Current Avg. Dice: 0.684297477497774


Validate (2584 / 10 Steps) (dice=0.70060): 100%|██████████| 17/17 [00:07<00:00,  2.13it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.7335874157793382 Current Avg. Dice: 0.6845810027683482


Validate (2652 / 10 Steps) (dice=0.66499): 100%|██████████| 17/17 [00:07<00:00,  2.15it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.7335874157793382 Current Avg. Dice: 0.6948514580726624


Validate (2788 / 10 Steps) (dice=0.66992): 100%|██████████| 17/17 [00:08<00:00,  2.11it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.7335874157793382 Current Avg. Dice: 0.6619636381373686


Validate (2856 / 10 Steps) (dice=0.69175): 100%|██████████| 17/17 [00:07<00:00,  2.17it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.7335874157793382 Current Avg. Dice: 0.6908275344792534


Validate (2992 / 10 Steps) (dice=0.70273): 100%|██████████| 17/17 [00:08<00:00,  2.10it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.7335874157793382 Current Avg. Dice: 0.7068013618974125


Validate (3060 / 10 Steps) (dice=0.72550): 100%|██████████| 17/17 [00:07<00:00,  2.14it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.7335874157793382 Current Avg. Dice: 0.7087247336612028


Validate (3196 / 10 Steps) (dice=0.70436): 100%|██████████| 17/17 [00:07<00:00,  2.14it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.7335874157793382 Current Avg. Dice: 0.7275888008229873


Validate (3264 / 10 Steps) (dice=0.71210): 100%|██████████| 17/17 [00:07<00:00,  2.14it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.7335874157793382 Current Avg. Dice: 0.6963957057279699


Validate (3400 / 10 Steps) (dice=0.72796): 100%|██████████| 17/17 [00:07<00:00,  2.23it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.7335874157793382 Current Avg. Dice: 0.7128325490390554


Validate (3468 / 10 Steps) (dice=0.74838): 100%|██████████| 17/17 [00:07<00:00,  2.15it/s]


Model Was Saved ! Current Best Avg. Dice: 0.7623843971420737 Current Avg. Dice: 0.7623843971420737


Validate (3536 / 10 Steps) (dice=0.75327): 100%|██████████| 17/17 [00:07<00:00,  2.21it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.7623843971420737 Current Avg. Dice: 0.7525574775303111


Validate (3672 / 10 Steps) (dice=0.78814): 100%|██████████| 17/17 [00:08<00:00,  2.08it/s]


Model Was Saved ! Current Best Avg. Dice: 0.7840510782073525 Current Avg. Dice: 0.7840510782073525


Validate (3740 / 10 Steps) (dice=0.78277): 100%|██████████| 17/17 [00:08<00:00,  2.06it/s]


Model Was Saved ! Current Best Avg. Dice: 0.8054105253780589 Current Avg. Dice: 0.8054105253780589


Validate (3876 / 10 Steps) (dice=0.83311): 100%|██████████| 17/17 [00:08<00:00,  2.10it/s]


Model Was Saved ! Current Best Avg. Dice: 0.8469615473466761 Current Avg. Dice: 0.8469615473466761


Validate (3944 / 10 Steps) (dice=0.81273): 100%|██████████| 17/17 [00:07<00:00,  2.15it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.8469615473466761 Current Avg. Dice: 0.839085620992324


Validate (4080 / 10 Steps) (dice=0.81674): 100%|██████████| 17/17 [00:08<00:00,  2.09it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.8469615473466761 Current Avg. Dice: 0.8236474009121165


Validate (4148 / 10 Steps) (dice=0.82067): 100%|██████████| 17/17 [00:07<00:00,  2.18it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.8469615473466761 Current Avg. Dice: 0.8300798500285429


Validate (4284 / 10 Steps) (dice=0.83219): 100%|██████████| 17/17 [00:08<00:00,  2.12it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.8469615473466761 Current Avg. Dice: 0.8375214338302612


Validate (4352 / 10 Steps) (dice=0.84441): 100%|██████████| 17/17 [00:07<00:00,  2.18it/s]


Model Was Saved ! Current Best Avg. Dice: 0.8613660405663883 Current Avg. Dice: 0.8613660405663883


Validate (4488 / 10 Steps) (dice=0.83119): 100%|██████████| 17/17 [00:07<00:00,  2.13it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.8613660405663883 Current Avg. Dice: 0.8482537865638733


Validate (4556 / 10 Steps) (dice=0.83632): 100%|██████████| 17/17 [00:07<00:00,  2.16it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.8613660405663883 Current Avg. Dice: 0.8486012290505802


Validate (4692 / 10 Steps) (dice=0.85008): 100%|██████████| 17/17 [00:08<00:00,  2.12it/s]


Model Was Saved ! Current Best Avg. Dice: 0.8796310284558464 Current Avg. Dice: 0.8796310284558464


Validate (4760 / 10 Steps) (dice=0.83760): 100%|██████████| 17/17 [00:07<00:00,  2.17it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.8796310284558464 Current Avg. Dice: 0.8645319623105666


Validate (4896 / 10 Steps) (dice=0.84658): 100%|██████████| 17/17 [00:08<00:00,  2.07it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.8796310284558464 Current Avg. Dice: 0.8554971568724689


Validate (4964 / 10 Steps) (dice=0.87662): 100%|██████████| 17/17 [00:07<00:00,  2.17it/s]


Model Was Saved ! Current Best Avg. Dice: 0.8917930757298189 Current Avg. Dice: 0.8917930757298189


Validate (5100 / 10 Steps) (dice=0.84880): 100%|██████████| 17/17 [00:08<00:00,  2.07it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.8917930757298189 Current Avg. Dice: 0.8615578342886532


Validate (5168 / 10 Steps) (dice=0.80306): 100%|██████████| 17/17 [00:08<00:00,  2.10it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.8917930757298189 Current Avg. Dice: 0.8146607034346637


Validate (5236 / 10 Steps) (dice=0.85590): 100%|██████████| 17/17 [00:07<00:00,  2.20it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.8917930757298189 Current Avg. Dice: 0.8731017778901493


Validate (5372 / 10 Steps) (dice=0.85177): 100%|██████████| 17/17 [00:08<00:00,  2.10it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.8917930757298189 Current Avg. Dice: 0.8707289871047524


Validate (5440 / 10 Steps) (dice=0.84127): 100%|██████████| 17/17 [00:07<00:00,  2.18it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.8917930757298189 Current Avg. Dice: 0.8602551291970646


Validate (5576 / 10 Steps) (dice=0.85401): 100%|██████████| 17/17 [00:07<00:00,  2.16it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.8917930757298189 Current Avg. Dice: 0.8808436533984016


Validate (5644 / 10 Steps) (dice=0.85535): 100%|██████████| 17/17 [00:07<00:00,  2.18it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.8917930757298189 Current Avg. Dice: 0.8694178777582505


Validate (5780 / 10 Steps) (dice=0.82920): 100%|██████████| 17/17 [00:08<00:00,  2.08it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.8917930757298189 Current Avg. Dice: 0.8458051295841441


Validate (5848 / 10 Steps) (dice=0.87628): 100%|██████████| 17/17 [00:08<00:00,  2.09it/s]


Model Was Saved ! Current Best Avg. Dice: 0.8945056725950802 Current Avg. Dice: 0.8945056725950802


Validate (5984 / 10 Steps) (dice=0.85269): 100%|██████████| 17/17 [00:08<00:00,  2.07it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.8945056725950802 Current Avg. Dice: 0.8707270131391638


Validate (6052 / 10 Steps) (dice=0.83868): 100%|██████████| 17/17 [00:07<00:00,  2.15it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.8945056725950802 Current Avg. Dice: 0.8613620014751658


Validate (6188 / 10 Steps) (dice=0.83943): 100%|██████████| 17/17 [00:08<00:00,  2.02it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.8945056725950802 Current Avg. Dice: 0.8663875039886025


Validate (6256 / 10 Steps) (dice=0.85699): 100%|██████████| 17/17 [00:07<00:00,  2.15it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.8945056725950802 Current Avg. Dice: 0.8764225665260764


Validate (6392 / 10 Steps) (dice=0.85501): 100%|██████████| 17/17 [00:07<00:00,  2.19it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.8945056725950802 Current Avg. Dice: 0.8725233253310708


Validate (6460 / 10 Steps) (dice=0.84536): 100%|██████████| 17/17 [00:08<00:00,  2.12it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.8945056725950802 Current Avg. Dice: 0.8608700317495009


Validate (6596 / 10 Steps) (dice=0.86098): 100%|██████████| 17/17 [00:07<00:00,  2.19it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.8945056725950802 Current Avg. Dice: 0.8777117869433235


Validate (6664 / 10 Steps) (dice=0.85464): 100%|██████████| 17/17 [00:08<00:00,  2.11it/s]


Model Was Not Saved ! Current Best Avg. Dice: 0.8945056725950802 Current Avg. Dice: 0.8798549035016228


Validate (6800 / 10 Steps) (dice=0.85903): 100%|██████████| 17/17 [00:08<00:00,  2.09it/s]


Model Was Saved ! Current Best Avg. Dice: 0.8974884713397306 Current Avg. Dice: 0.8974884713397306


Training (6879 / 10000 Steps) (loss=0.65215):  18%|█▊        | 12/68 [00:07<00:34,  1.64it/s]


RuntimeError: ignored

#5000 Steps
Fold1 - 0.9086 Fold2 - 0.87 Fold3 - 0.895 Fold4 - 0.888 Fold5 - 87.19
#10000 Steps
Fold1 - 0.9379 Fold2 - 0.9170 Fold3-

In [ ]:
plt.figure(1, (12, 6))
plt.subplot(1, 2, 1)
plt.title("Iteration Average Loss")
x = [eval_num * (i + 1) for i in range(len(epoch_loss_values))]
y = epoch_loss_values
plt.xlabel("Iteration")
plt.plot(x, y)
plt.grid()
plt.subplot(1, 2, 2)
plt.title("Val Mean Dice")
x = [eval_num * (i + 1) for i in range(len(metric_values))]
y = metric_values
plt.xlabel("Iteration")
plt.plot(x, y)
plt.grid()
plt.show()

In [ ]:
from glob import glob
original_data_folder = '/content/drive/MyDrive/template/our-git-project/preprocessed_images/'
mask_data_folder = '/content/drive/MyDrive/template/our-git-project/preprocessed_images/'

# List of image file paths
vol  = sorted(glob(os.path.join(original_data_folder, '*_orig.nii.gz')))
segs  = sorted(glob(os.path.join(mask_data_folder, '*_masks.nii.gz')))

pixdim=(1.0, 1.0, 1.0)
a_min=-200
a_max=200
spatial_size=[150,150,150]

data_dicts = [{"vol": image_name, "seg": label_name} for image_name, label_name in zip(vol, segs)]
train_files, test_files = data_dicts[:-9], data_dicts[-9:]


test_transforms = Compose(
        [
            LoadImaged(keys=["vol", "seg"]),
            EnsureChannelFirstd(keys=["vol", "seg"]),
            #Spacingd(keys=["vol", "seg"], pixdim=pixdim, mode=("bilinear", "nearest")),
            Orientationd(keys=["vol", "seg"], axcodes="RAS"),
            #NormalizeIntensityd(keys="vol", nonzero=True, channel_wise=True),
            CropForegroundd(keys=['vol', 'seg'], source_key='vol'),
            RandCropByPosNegLabeld(
            keys=["vol", "seg"],
            label_key="seg",
            spatial_size=(64, 64, 64),
            pos=4,
            neg=1,
            num_samples=5,
            image_key="vol",
            image_threshold=0,),
            RandFlipd(
            keys=["vol", "seg"],
            spatial_axis=[0],
            prob=0.5,
        ),
            RandFlipd(
            keys=["vol", "seg"],
            spatial_axis=[0],
            prob=0.5,
        ),
            RandFlipd(
            keys=["vol", "seg"],
            spatial_axis=[1],
            prob=0.5,
        ),

        RandFlipd(
            keys=["vol", "seg"],
            spatial_axis=[2],
            prob=0.5,
        ),
        RandRotate90d(
            keys=["vol", "seg"],
            prob=0.5,
            max_k=3,
        ),
        ToTensord(keys=["vol", "seg"]),


        ]
    )

test_ds = CacheDataset(data=test_files, transform=test_transforms, cache_rate=1.0)
test_loader = DataLoader(test_ds, batch_size=1)

In [ ]:
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

model = UNETR(
    in_channels=1,
    out_channels=2,
    img_size=(64, 64, 64),
    feature_size=16,
    hidden_size=768,
    mlp_dim=3072,
    num_heads=12,
    pos_embed="conv",
    norm_name="instance",
    res_block=True,
    dropout_rate=0.0,
).to(device)


In [ ]:
model.load_state_dict(torch.load(
    os.path.join(model_dir, "best_metric_model.pth")))
model.eval()

In [ ]:
from monai.utils import first
sw_batch_size = 4
roi_size = (64, 64, 64)
with torch.no_grad():
    test_patient = first(test_loader)
    t_volume = test_patient['vol'].cuda()
    val_labels = test_patient['seg'].cuda()

    val_outputs = sliding_window_inference(t_volume.to(device), roi_size, sw_batch_size, model)
    val_labels_list = decollate_batch(val_labels)
    val_labels_convert = [post_label(val_label_tensor) for val_label_tensor in val_labels_list]
    val_outputs_list = decollate_batch(val_outputs)
    val_output_convert = [post_pred(val_pred_tensor) for val_pred_tensor in val_outputs_list]
    dice_metric(y_pred=val_output_convert, y=val_labels_convert)
    dice = dice_metric.aggregate().item()
    print("Test Dice Value: ",dice)

In [ ]:
sw_batch_size = 20
roi_size = (64, 64, 64)
with torch.no_grad():
    test_patient = first(test_loader)
    t_volume = test_patient['vol']
    #t_segmentation = test_patient['seg']

    test_outputs = sliding_window_inference(t_volume.to(device), roi_size, sw_batch_size, model)
    sigmoid_activation = Activations(sigmoid=True)
    test_outputs = sigmoid_activation(test_outputs)
    test_outputs = test_outputs > 0.69

    for i in range(31):
        # plot the slice [:, :, 80]
        plt.figure("check", (18, 6))
        plt.subplot(1, 3, 1)
        plt.title(f"image {i}")
        plt.imshow(test_patient["vol"][0, 0, :, :, i], cmap="gray")
        plt.subplot(1, 3, 2)
        plt.title(f"label {i}")
        plt.imshow(test_patient["seg"][0, 0, :, :, i] != 0)
        plt.subplot(1, 3, 3)
        plt.title(f"output {i}")
        plt.imshow(test_outputs.detach().cpu()[0, 1, :, :, i])
        plt.show()

In [ ]:
from monai.utils import set_determinism
from monai.transforms import (
    AsDiscrete,
    EnsureChannelFirstd,
    Compose,
    CropForegroundd,
    KeepLargestConnectedComponent,
    LabelToContour,
    LoadImaged,
    Orientationd,
    RandCropByPosNegLabeld,
    ScaleIntensityRanged,
    Spacingd,
    EnsureTyped,
    EnsureType,
    RemoveSmallObjects,

)
from monai.networks.nets import UNet
from monai.networks.layers import Norm
from monai.metrics import DiceMetric
from monai.losses import DiceCELoss
from monai.inferers import sliding_window_inference
from monai.data import CacheDataset, DataLoader, decollate_batch
from monai.config import print_config
from monai.apps import download_and_extract
import torch
import matplotlib.pyplot as plt
import tempfile
import shutil
import os
import glob


print_config()

In [ ]:
model.load_state_dict(torch.load(
    os.path.join(model_dir, "best_metric_model.pth")))
model.eval()
with torch.no_grad():
        test_patient = first(test_loader)
        t_volume = test_patient['vol']
        #t_segmentation = test_patient['seg']

        test_outputs = sliding_window_inference(t_volume.to(device), roi_size, sw_batch_size, model)
        sigmoid_activation = Activations(sigmoid=True)
        test_outputs = sigmoid_activation(test_outputs)
        test_outputs_1 = test_outputs > 0.69

        for i in range(51):
            # plot the slice [:, :, 80]
            plt.figure("check", (20, 4))
            plt.subplot(1, 7, 1)
            plt.title(f"image {i}")
            plt.imshow(test_patient['vol'].detach().cpu()[0, 0, :, :, i], cmap="gray")
            plt.subplot(1, 7, 2)
            plt.title(f"argmax {i}")
            argmax = [AsDiscrete(argmax=True)(j) for j in decollate_batch(test_outputs)]
            plt.imshow(argmax[0].detach().cpu()[0, :, :, i])
            plt.subplot(1, 7, 3)
            plt.title(f"largest {i}")
            largest = [KeepLargestConnectedComponent(applied_labels=[1])(j) for j in argmax]
            plt.imshow(largest[0].detach().cpu()[0, :, :, i])
            plt.subplot(1, 7, 4)
            plt.title(f"contour {i}")
            contour = [LabelToContour()(j) for j in largest]
            plt.imshow(contour[0].detach().cpu()[0, :, :, i])
            plt.subplot(1, 7, 5)
            plt.title(f"map image {i}")
            map_image = contour[0] + t_volume.to(device)[0]
            plt.imshow(map_image.detach().cpu()[0, :, :, i], cmap="gray")
            plt.subplot(1, 7, 6)
            plt.title(f"output {i}")
            plt.imshow(test_outputs_1.detach().cpu()[0, 1, :, :, i])
            plt.subplot(1, 7, 7)
            plt.title(f"label {i}")
            plt.imshow(test_patient["seg"][0, 0, :, :, i] != 0)
            plt.show()

In [ ]:
model.load_state_dict(torch.load(
    os.path.join(model_dir, "best_metric_model.pth")))
model.eval()
with torch.no_grad():
    for i, val_data in enumerate(test_loader):
        roi_size = (64, 64, 64)
        sw_batch_size = 4
        val_data = val_data["vol"].to(device)
        val_output = sliding_window_inference(val_data, roi_size, sw_batch_size, model)
        # plot the slice [:, :, 80]
        plt.figure("check", (20, 4))
        plt.subplot(1, 5, 1)
        plt.title(f"image {i}")
        plt.imshow(val_data.detach().cpu()[0, 0, :, :, 30], cmap="gray")
        plt.subplot(1, 5, 2)
        plt.title(f"argmax {i}")

        argmax = [AsDiscrete(argmax=True)(i) for i in decollate_batch(val_output)]
        plt.imshow(argmax[0].detach().cpu()[0, :, :, 30])
        plt.subplot(1, 5, 3)
        plt.title(f"largest {i}")
        largest = [KeepLargestConnectedComponent(applied_labels=[1])(i) for i in argmax]
        plt.imshow(largest[0].detach().cpu()[0, :, :, 30])
        plt.subplot(1, 5, 4)
        plt.title(f"contour {i}")
        contour = [LabelToContour()(i) for i in largest]
        plt.imshow(contour[0].detach().cpu()[0, :, :, 30])
        plt.subplot(1, 5, 5)
        plt.title(f"map image {i}")
        map_image = contour[0] + val_data[0]
        plt.imshow(map_image.detach().cpu()[0, :, :, 30], cmap="gray")

        plt.show()